crates/noirc_evaluator/src/ssa/code_gen.rs

```diff
@@ -0,0 +1,1006 @@
+        let i = node::Instruction::new(
+            node::Operation::cast,
+            lhs,
+            lhs,
+            rtype,
+            Some(self.current_block),
+        );
+        self.add_object(node::NodeObj::Instr(i))
+    }
+
+    fn evaluate_infix_expression(
+        &mut self,
+        lhs: arena::Index,
+        rhs: arena::Index,
+        op: HirBinaryOp,
+    ) -> Result<arena::Index, RuntimeError> {
+        let ltype = self.get_object_type(lhs);
+
+        let optype = ltype;
+        //TODO if type differs try to cast them: check how rust handle this
+        //let rtype = self.get_object_type(rhs);
+        //and else returns an error
+
+        // Get the opcode from the infix operator
+        let opcode = node::to_operation(op.kind, optype);
+        //TODO we should validate the types with the opcode
+        Ok(self.new_instruction(lhs, rhs, opcode, optype))
+    }
+
+    pub fn evaluate_statement(
+        &mut self,
+        env: &mut Environment,
+        stmt_id: &StmtId,
+    ) -> Result<arena::Index, RuntimeError> {
+        let statement = self.context().def_interner.statement(stmt_id);
+        match statement {
+            HirStatement::Private(x) => self.handle_private_statement(env, x),
+            HirStatement::Constrain(constrain_stmt) => {
+                self.handle_constrain_statement(env, constrain_stmt)
+            }
+            HirStatement::Const(x) => {
+                //let variable_name: String = self.context().def_interner.ident_name(&x.identifier);
+                // const can only be integers/Field elements, cannot involve the witness, so we can possibly move this to
+                // analysis. Right now it would not make a difference, since we are not compiling to an intermediate Noir format
+                //let span = self.context().def_interner.expr_span(&x.expression);
+                //TODO the result of expression_to_object should be an assignement, we should modify the lhs to specify it is a const
+                // and then forbid any other assignement with the same variable during the SSA phase (and instead of applying the SSA form of it).
+                self.expression_to_object(env, &x.expression)
+            }
+            HirStatement::Expression(expr) | HirStatement::Semi(expr) => {
+                self.expression_to_object(env, &expr)
+            }
+            HirStatement::Let(let_stmt) => {
+                // let statements are used to declare a higher level object
+                self.handle_let_statement(env, let_stmt)
+            }
+            HirStatement::Assign(assign_stmt) => {
+                //left hand is already declared
+                //TODO clarify what is an HirStatement::Assign vs HirStatement::Let vs HirStatement::Priv
+                //e.g 'let a=x+2;' could be let:(assign:(a, x+2)) or assign:(let a, x+2)
+                //For now we do all the job here but it should be splitted
```

Comment:
`HirStatement::Assign` is a reassignment like in `a = 2;` where `HirStatement::Let` is a definition as in `let a = 2;` so they're mutually exclusive.
Aside: I think the "assign" node is currently a lie. It implies mutability in other languages but from my understanding is only meant to represent shadowing in noir.

Reply:
Thank you for the clarification; I remove the variable creation from the assign.
Note that assign is a real assign in Noir.

Reply:
I added it back, because I do not parse arguments of main. I should modify the parseABI method.